TrajectoryExecutor interface

lgw903 opened the pull request:
This PR contains: 

- TrajectoryExecutor
- TrajectoryResult
- KinematicSimulationTrajectoryExecutor
- Tests for KinematicSimulationTrajectoryExecutor.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/personalrobotics/aikido/96)
<!-- Reviewable:end -->

mkoval commented:
This is a good start. I have a few high-level comments:

- There are some concurrency issues in `KinematicSimulationTrajectoryExecutor`.
- I am no longer sure whether the empty `TrajectoryResult` base class is the best way of achieving the extensibility that we're looking for.
- We need to have a bigger discussion on how to interpret `getStartTime` on `Trajectory`s. I suggest punting on this for now to avoid putting this pull request in limbo. I'm sure @psigen and @cdellin will have thoughts on this.

---

Reviewed 9 of 10 files at r2.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp, line 13 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGc5DpokbLyhOlBFa-T:-KGc5DpokbLyhOlBFa-U:-994220736) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp#L13)):*
Don't `virtual`ly inherit unless you need to.

---

*[aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp, line 17 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdJMtunob4CB53U2nY:-KGdJMtunob4CB53U2nZ:941670069) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp#L17)):*
Take `_skeleton` by value and `std::move` it internally into a member variable.

---

*[aikido/include/aikido/control/TrajectoryExecutor.hpp, line 17 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdd7QRQuRVrsoa-2u5:-KGdd7QRQuRVrsoa-2u6:-1408735128) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/include/aikido/control/TrajectoryExecutor.hpp#L17)):*
You should specifically document why return `TrajectoryResultPtr`, a `std::shared_ptr` to an empty type, instead of `void`. This is not obvious - I had to think about it for a few minutes to recall the reasoning behind this.

I'm also beginning to question whether this is the right decision. It would be simpler to make this return a `std::future<void>` and add additional `execute` overloads on sub-classes of `TrajectoryExecutor` that provide additional functionality. For example, the `ros_control` executor could have a `executeWithFeedback` function that uses the actionlib feedback messages to reconstruct an Aikido `Trajectory` from the path *actually followed by the robot*.

@ClintLiddick @lgw903: Thoughts?

---

*[aikido/src/CMakeLists.txt, line 10 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdIyeeCFzI_d0nEtnP:-KGdIyeeCFzI_d0nEtnQ:-685235722) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/CMakeLists.txt#L10)):*
We don't generally list external dependencies (e.g. `dart`) here. These comments are mostly intended to make sure that an `add_subdirectory` call occurs *after* the `add_subdirectory` calls for the components that it depends on.

Also, `python` should be last, since it depends on *everything*.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 67 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdg1Eyo3JRBcqW8pFc:-KGdg1Eyo3JRBcqW8pFd:258408857) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L67)):*
Include the DOF name in the error message for debugging purposes.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 68 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdJ_eD-eFmj2LX1qAN:-KGdJ_eD-eFmj2LX1qAO:2002953830) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L68)):*
We lock and unlock `mMutex` a few times here. This could lead to concurrency problems. For example, the following could happen if the user calls `execute` in two threads:

- Thread 1: checks `mTraj`
- Thread 2: checks `mTraj`
- Thread 1: starts executing `traj1`
- Thread 2: starts executing `traj2`

I suggest doing *all* error checking on `_traj` before locking `mMutex`, then modifying all member variables (e.g. `mTraj` and `mPromise`) under *the same lock*.

Also, you should lock `mSkeleton->getMutex()` before accessing any attributes on `mSkeleton` (e.g. `getDof`).

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 87 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdg9LjJZD9xCuXRELI:-KGdg9LjJZD9xCuXRELJ:1011572409) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L87)):*
Make this a parameter.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 93 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdgFGzGo8Cu1A50Tsp:-KGdgFGzGo8Cu1A50Tsq:-1183155992) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L93)):*
You can remove this local variable by simply moving the `if(!mRunning) break;` check here.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 100 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdgLqTbeQDBwpbTfx1:-KGdgLqUssCeGc0ivhPk:-602666845) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L100)):*
It's unsafe to access `mTraj` unless the lock is acquired.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 104 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdgWmAoedjk-sOSMH-:-KGdgWmAoedjk-sOSMH0:-1451788487) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L104)):*
This is essentially a spinlock. This is generally a bad idea because it can waste a lot of CPU time. I suggest making one of two changes:

1. use a `std::condition_variable` to trigger this event
2. sleep for the usual `period` even if no trajectory is executing

I slightly prefer (1), but (2) would also be acceptable.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 116 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdhEdVflGSkd0cr7mV:-KGdhEdVflGSkd0cr7mW:780040851) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L116)):*
I am nervous about putting this in a second nested loop. This makes it impossible to interrupt a running trajectory, even if this class is destructed. Instead, I would prefer to have *one loop* that runs at a fixed `period` and processes the currently-running trajectory. Alternatively, you could also check `mRunning` in the inner loop as well.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 122 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdhaq66d-6IEWMQCX4:-KGdhaq66d-6IEWMQCX5:-1519542021) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L122)):*
It's not clear what `getStartTime()` is relative to, so I don't think this is the correct behavior. I suggest ignoring `getStartTime()` entirely for the sake of this pull request and treating trajectories as if they run from `0.0` to `traj->getDuration()`.

We should have a separate discussion about how to handle non-zero start times. Could you create a Github issue for that?

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 146 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdio9oapW-xuwdozGr:-KGdio9oapW-xuwdozGs:-1291519940) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L146)):*
We should `set_exception` on `mPromise` if execution terminates while a trajectory is running.

---

*[aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp, line 94 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdjyQD9BxpSSHMnEhW:-KGdjyQD9BxpSSHMnEhX:924900234) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp#L94)):*
Add a test that tries to execute a trajectory that does *not* have a `MetaSkeletonStateSpace`.

---

*[aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp, line 96 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdjKukK9eYGe_u2Ebc:-KGdjKulWGKGWkOigRBm:-808970) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp#L96)):*
Give this test a more descriptive name. Maybe something like this: `execute_GetOnFuture_TrajectoryWasExecuted`.

---

*[aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp, line 111 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdjjFPV-xWxwhCCoC6:-KGdjjFQVcihTlXBFeyU:548567605) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp#L111)):*
A better name would be: `execute_TrajectoryIsAlreadyRunning_Throws`.

---

*[aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp, line 130 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdjoOzx-G17Rm41ILo:-KGdjoOzx-G17Rm41ILp:1262978623) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp#L130)):*
A better name would be `execute_TrajectoryFinished_DoesNotThrow`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-:-KGdkCC_cvzSCEJRCbFU:1995404932)*
<!-- Sent from Reviewable.io -->

lgw903 commented:
@mkoval 
Thanks for the feedback! 
Comments addressed in the following way:
- constructor takes `mPeriod`
- uses `condition_variable` to `wait` on `mRunning` and `mTraj`.
- cleaned up lock usage. 
- added more tests, changed names

---

Review status: 3 of 9 files reviewed at latest revision, 17 unresolved discussions.

---

*[aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp, line 13 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGc5DpokbLyhOlBFa-T:-KH0RQ4rRHOFSOX3NaXZ:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp#L13)):*
Done.

---

*[aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp, line 17 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdJMtunob4CB53U2nY:-KH0RYr0lkvoiiyKZ_yx:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp#L17)):*
Done.

---

*[aikido/include/aikido/control/TrajectoryExecutor.hpp, line 17 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdd7QRQuRVrsoa-2u5:-KH0RoeNuq8mgnjkoZDa:2005496156) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/include/aikido/control/TrajectoryExecutor.hpp#L17)):*
I like that idea. That would reduce having to return a `std::shared_ptr` to an empty type in many instances, and we may not use the feedback function in many cases. I will go ahead and make that change, unless @ClintLiddick has other thoughts? 

---

*[aikido/src/CMakeLists.txt, line 10 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdIyeeCFzI_d0nEtnP:-KH0T_jyvZEHLEMYO3J4:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/CMakeLists.txt#L10)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 67 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdg1Eyo3JRBcqW8pFc:-KH0VPO2i7ze4mWp9tp5:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L67)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 93 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdgFGzGo8Cu1A50Tsp:-KH0XvxSQgVkQXW85ZJl:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L93)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 100 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdgLqTbeQDBwpbTfx1:-KH39Dk6_Ka4amdiG6-K:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L100)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 104 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdgWmAoedjk-sOSMH-:-KH39CqelxKsIuXHHFIL:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L104)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 116 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdhEdVflGSkd0cr7mV:-KH39CBJ_aESNINC_UbY:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L116)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 122 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdhaq66d-6IEWMQCX4:-KH39BMZjY_fggDzldRN:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L122)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 146 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdio9oapW-xuwdozGr:-KH38huvRrVUwDh4ru58:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L146)):*
Done.

---

*[aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp, line 94 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdjyQD9BxpSSHMnEhW:-KH3AReJIndFY2sYdEow:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp#L94)):*
Done.

---

*[aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp, line 96 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdjKukK9eYGe_u2Ebc:-KH3ASBBPZmcNfXdLswd:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp#L96)):*
Done.

---

*[aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp, line 111 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdjjFPV-xWxwhCCoC6:-KH3AaqW2CeaQCN2txJI:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp#L111)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-:-KH3BHsyrLfeLTQTi3ev:-897889932)*
<!-- Sent from Reviewable.io -->

lgw903 commented:
Review status: 3 of 9 files reviewed at latest revision, 17 unresolved discussions.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 68 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdJ_eD-eFmj2LX1qAN:-KH3BpBnq8k4W-pFPG0f:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L68)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 87 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdg9LjJZD9xCuXRELI:-KH3BqbBZFo1KasVR9Mp:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L87)):*
Done.

---

*[aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp, line 130 \[r2\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KGdjoOzx-G17Rm41ILo:-KH3BrQmMINFOJ2djFDh:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/a3a2389b2eb86582afba406bf085f3cd245ad8a8/aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp#L130)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/personalrobotics/aikido/96)*
<!-- Sent from Reviewable.io -->

mkoval commented:
Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp, line 31 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5WyTLYumE23swbYPa:-KH5WyTLYumE23swbYPb:-1722952765) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp#L31)):*
Either add documentation to `execute` or add "Documentation inherited."

---

*[aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp, line 39 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5Wj28L0PHmMqSo3Xh:-KH5Wj28L0PHmMqSo3Xi:-2068417569) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp#L39)):*
Nit: There's no harm in making the `private` members use Doxygen-style (`///`) comments.

---

*[aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp, line 47 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5WUdnXyfDD7wOp5CK:-KH5WUdnXyfDD7wOp5CL:-1584208953) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp#L47)):*
Nit: the name `mSpinLock` is misleading because `std::mutex` is not a [spinlock](https://en.wikipedia.org/wiki/Spinlock) on most architectures.

---

*[aikido/include/aikido/control/TrajectoryExecutor.hpp, line 17 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5XMRGD-Q9MgcihUB3:-KH5XMRGD-Q9MgcihUB4:764994276) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/include/aikido/control/TrajectoryExecutor.hpp#L17)):*
Missing docstring.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 31 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5YKooWRr6ORpK6ES1:-KH5YKooWRr6ORpK6ES2:-969360430) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L31)):*
I believe an `std::duration` can be negative, so you should also check for that case. You can check for that by testing `mPeriod.counts() > 0`.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 61 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5Z3mF4cHciBuVNJ_d:-KH5Z3mF4cHciBuVNJ_e:1633667037) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L61)):*
Missing a space between "Executor's" and "MetaSkeletonStateSpace" in this message.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 71 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5ZBc8xtMbqdq-RoJe:-KH5ZBc8xtMbqdq-RoJf:1711095317) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L71)):*
You should acquire this lock once, outside of the loop. This is slow (we must acquire the lock many times) and dangerous (the `Joint`s and `DegreeOfFreedom`s in the `Skeleton` could change as we're iterating over it).

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 134 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5_TmOTe2Gq7xAHBGo:-KH5_TmOTe2Gq7xAHBGp:325979158) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L134)):*
`lockSpin` will automatically be unlocked (by RAII) when you exit the loop.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 139 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5_aLxjx32aOdztqun:-KH5_aLxjx32aOdztquo:-244361290) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L139)):*
We already know that `getStateSpace()` is a `MetaSkeletonStateSpace` from the check in `execute`. I suggest replacing the `dynamic_pointer_cast` with a `static_pointer_cast` for efficiency. It's also pointless to use `dynamic_pointer_cast` without checking the return value, because we'll just SEGFAULT on the next line. :smile:

---

*[aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp, line 88 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5aEI3bhWRf-uHgz6y:-KH5aEI4Tn7GVrxMoM1e:32717778) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp#L88)):*
Also test for a negative period.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/personalrobotics/aikido/96)*
<!-- Sent from Reviewable.io -->

lgw903 commented:
Review status: 5 of 9 files reviewed at latest revision, 10 unresolved discussions.

---

*[aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp, line 31 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5WyTLYumE23swbYPa:-KH6t2uJpDdnlsfzMqmM:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp#L31)):*
Done.

---

*[aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp, line 39 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5Wj28L0PHmMqSo3Xh:-KH6t3ZOe2CmLJOJ0Rn2:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp#L39)):*
Done.

---

*[aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp, line 47 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5WUdnXyfDD7wOp5CK:-KH6tEkc6MJDK0RaEVPR:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/include/aikido/control/KinematicSimulationTrajectoryExecutor.hpp#L47)):*
Done.

---

*[aikido/include/aikido/control/TrajectoryExecutor.hpp, line 17 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5XMRGD-Q9MgcihUB3:-KH6un-_fhF56JnIQwF_:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/include/aikido/control/TrajectoryExecutor.hpp#L17)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 31 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5YKooWRr6ORpK6ES1:-KH6v7eDTdUCbNm-gEEJ:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L31)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 61 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5Z3mF4cHciBuVNJ_d:-KH6vANY0odxafpyy3O3:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L61)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 71 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5ZBc8xtMbqdq-RoJe:-KH6w71zuy3ASDQaCl04:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L71)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 134 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5_TmOTe2Gq7xAHBGo:-KH6wvf4H9gFrRB6kV9W:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L134)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 139 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5_aLxjx32aOdztqun:-KH6wv8M9AQijZCMAdoP:-1743841245) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L139)):*
Can't do this because `MetaSkeletonStateSpace` inherits `CartesianProduct` which virtually inherits `StateSpace`. 

---

*[aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp, line 88 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5aEI3bhWRf-uHgz6y:-KH6wt5piOBLs2fBlLgz:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/tests/control/test_KinematicSimulationTrajectoryExecutor.cpp#L88)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/personalrobotics/aikido/96)*
<!-- Sent from Reviewable.io -->

mkoval commented:
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[aikido/include/aikido/control/TrajectoryExecutor.hpp, line 17 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5XMRGD-Q9MgcihUB3:-KH727mdcmmPfMxnb9gD:1178238793) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/include/aikido/control/TrajectoryExecutor.hpp#L17)):*
Missing a `\return` statement.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 31 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5YKooWRr6ORpK6ES1:-KH72C1ndPZoNy54qEse:377153143) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L31)):*
This should be `<= 0`, not `< 0`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/personalrobotics/aikido/96)*
<!-- Sent from Reviewable.io -->

lgw903 commented:
Review status: 6 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[aikido/include/aikido/control/TrajectoryExecutor.hpp, line 17 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5XMRGD-Q9MgcihUB3:-KH72fJNf1do4-kgrU95:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/include/aikido/control/TrajectoryExecutor.hpp#L17)):*
Done.

---

*[aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp, line 31 \[r4\]](https://reviewable.io:443/reviews/personalrobotics/aikido/96#-KH5YKooWRr6ORpK6ES1:-KH733aldmLSL8u-vCfg:-499146009) ([raw file](https://github.com/personalrobotics/aikido/blob/da391df5ade80215b6959aea13418bc743a36658/aikido/src/control/KinematicSimulationTrajectoryExecutor.cpp#L31)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/personalrobotics/aikido/96)*
<!-- Sent from Reviewable.io -->

mkoval commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/personalrobotics/aikido/96)*
<!-- Sent from Reviewable.io -->
